### PR TITLE
Issue/#145 s3 storing service

### PR DESCRIPTION
**Pull request by razine-bensari**
# Related Issue
- #145 

# Proposed changes
- added implementation for profil pic persistence using amazon s3

I mocked the tests. @abdulmansour please take over the comments on the PR.

# Checklist (if applicable)
- [DONE ] Tests

# Reviewer(s)
 - @DPigeon @abdulmansour

### CODE REVIEW

**Comment by razine-bensari** [comment]
CI is failaing because of ENVIRONMENT KEYS from amazon that are not set up :( @DPigeon  or @wltrf  can anyone of you set up the secrets in the CI

**Comment by DPigeon** [comment]
> CI is failaing because of ENVIRONMENT KEYS from amazon that are not set up :( @DPigeon or @wltrf can anyone of you set up the secrets in the CI

The keys have been setup for a while in the repo as secrets. Did you

> CI is failaing because of ENVIRONMENT KEYS from amazon that are not set up :( @DPigeon or @wltrf can anyone of you set up the secrets in the CI

The keys are already added. You probably forgot to change the CI workflow and include those in there.

**Comment by razine-bensari** [comment]
@DPigeon 

I added the secrets in the ci.yml file
:)

File: backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy [hunk 1]
```diff
@@ -235,4 +240,114 @@ class UserControllerTest extends Specification {
...
+      String firstName = "Billy"
+      String lastName = "Bob"
+      String alpacaApiKey = "RYFERH6ET5etETGTE6"
+      User user = createUser(email, username, password, firstName, lastName, alpacaApiKey)
+
+      and: "mock the database with the same user already registered"
+      userDao.findUserById(user.getId()) >> user
+
+      and: "A valid multipartFile is passed"
+      MultipartFile imageFile = Mock()
```
Comment (DPigeon) [inline backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy:264]: Minor comment: should we use a real image to test here? Or is this going to be done in #165 ?

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        catch (Exception e) {
+            throw new ExceptionAmazonS3("Failed to convert file");
+        }
+        return convFile;
+    }
+
+    private void cleanUpLocalFileCreated(File image) {
+        //delete the local image
+        if (!image.delete())
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:93]: Add some brackets here.

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            throw new ExceptionAmazonS3("Failed when deleting local file");
+    }
+
+    public void deleteImageFroms3Bucket(String bucketName, String url) {
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:97]: Change this name for `deleteImageFromS3Bucket` by making the s capitalized.

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        else {
+            throw new ExceptionAmazonS3("Invalid file format");
+        }
+    }
+
+    //need to convert MultipartFile => File
+    private File convertMultiPartToFile(MultipartFile file) {
+        File convFile;
+        try {
+            convFile = new File(file.getOriginalFilename());
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:79]: The multipartFile might be null here. I think it should be good if we wrap this up with a `requireNonNull` object. We could also do a test for that and see what happens when multipartfile is null.

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+            //upload and wait for completion
+            Upload upload = tm.upload(bucketName, keyName, image);
+            try {
+                upload.waitForCompletion();
+                this.cleanUpLocalFileCreated(image);
+                LOG.info("Uploaded {} successfully!", keyName);
+
+                //return image s3 url
+                String url = "https://" + bucketName + ".s3.amazonaws.com/" + keyName;
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:64]: Remove the String url line and simple return it like this to remove redundancy:
`return "https://" + bucketName + ".s3.amazonaws.com/" + keyName;`

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.util.Arrays;
+import java.util.List;
+
+@Component
+public class AmazonS3Facade {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AmazonS3Facade.class);
+    private AmazonS3 s3;
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:31]: I think this should be a final variable.

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                    .withMultipartUploadThreshold((long) (5 * 1024 * 1025))
+                    .build();
+
+            // generate unique name
+            String keyName = RandomStringUtils.randomAlphanumeric(17).toUpperCase() + "-" + image.getName();
+
+            //upload and wait for completion
+            Upload upload = tm.upload(bucketName, keyName, image);
+            try {
+                upload.waitForCompletion();
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:59]: Is this a blocking thread method? What happens if the upload never gets completed? Just curious.

File: backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java [hunk 8]
```diff
@@ -30,12 +34,17 @@
    private final ValidatorFactory validatorFactory;
    private final MoneyTreePasswordEncryption passwordEncryption;
    private static final Logger LOG = LoggerFactory.getLogger(DefaultUserService.class);
+   private final AmazonS3Service amazonS3Service;
+
+   @Value("${aws.profile.pictures.bucket}")
+   private String bucketName;
```
Comment (DPigeon) [inline backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java:40]: I was never able to inject @ Values like that. You'll have to show it to me. It was always null for me. Nice!

**Review by DPigeon — CHANGES_REQUESTED** [review]
Nice work! I left some comments. Fix the SonarCloud alerts too.

File: backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java (hunk 8 above)
Comment (wltrf) [reply backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java:40]: Maybe you were trying to inject inside one of the test, in which case null is normal. I'll explain: @value is a spring annotation and only works if the class is run with the spring runner class. In the test, it is currently the junit runner class that runs it and therefore it ignores (doesn't know) the spring annotation. FYI I changed it and will be merged when #165 is done.

File: backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy (hunk 1 above)
Comment (wltrf) [reply backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy:264]: This is not in scope for #165. #165 is about using the test db, unrelated to s3.

File: backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java (hunk 8 above)
Comment (DPigeon) [reply backend/src/main/java/com/capstone/moneytree/service/impl/DefaultUserService.java:40]: That might've been the case. Thanks!

**Comment by razine-bensari** [comment]
@DPigeon  @wltrf  thks for your comment, Ill take over the PR commetns and address your comments and sonar cloud issues 👍

File: backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy (hunk 1 above)
Comment (razine-bensari) [reply backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy:264]: I am mocking it for now because I wanted to tests the error paths.

The reason why I did not do an integration test is because of s3 limitations. I dont want to endup like we did with IEXCloud

File: backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java (hunk 7 above)
Comment (razine-bensari) [reply backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:59]: It is a blocking thread method. If it is not complete within a time deadline, it throws a InterruptedException

Comment (DPigeon) [reply backend/src/main/java/com/capstone/moneytree/facade/AmazonS3Facade.java:59]: Ok cool!

**Review by DPigeon — COMMENTED** [review]
Nice! There is still one code smell in the backend on SonarCloud to fix (about a removable unused import). I'll approve after. Otherwise it looks good.

File: backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy [hunk 9]
```diff
@@ -235,4 +247,267 @@ class UserControllerTest extends Specification {
       then: "Should throw CredentialNotFoundException"
       thrown(CredentialNotFoundException)
    }
+
+//   @Test
```
Comment (DPigeon) [inline backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy:251]: Were these test commented out because they were failing?

Comment (razine-bensari) [reply backend/src/test/groovy/com/capstone/moneytree/controller/UserControllerTest.groovy:251]: No this was the old test, I updated all new test below with more test cases and coverage, I need to remove this commented code, I will update the PR. Cant believe I missed that

**Review by DPigeon — APPROVED** [review]
Good job. All good.
